lib/gshadow.c

```diff
@@ -155,29 +155,26 @@ void endsgent (void)
 		return NULL;
 	}
 
-	if (fgetsx(buf, buflen, fp) == buf) {
-		while (   ((cp = strrchr (buf, '\n')) == NULL)
-		       && (feof (fp) == 0)) {
-			size_t len;
-
-			cp = REALLOC(buf, buflen * 2, char);
-			if (NULL == cp) {
-				return NULL;
-			}
-			buf = cp;
-			buflen *= 2;
-
-			len = strlen (buf);
-			if (fgetsx (&buf[len],
-			            (int) (buflen - len),
-			            fp) != &buf[len]) {
-				return NULL;
-			}
+	if (fgets(buf, buflen, fp) == NULL)
+		return NULL;
+
+	while (   (strrchr(buf, '\n') == NULL)
+	       && (feof (fp) == 0)) {
+		size_t len;
+
+		cp = REALLOC(buf, buflen * 2, char);
+		if (NULL == cp) {
+			return NULL;
 		}
-		stpsep(buf, "\n");
-		return (sgetsgent (buf));
+		buf = cp;
+		buflen *= 2;
+
+		len = strlen (buf);
+		if (fgets(&buf[len], buflen - len, fp) == NULL)
+			return NULL;
 	}
-	return NULL;
+	stpsep(buf, "\n");
+	return (sgetsgent (buf));
```

Comment:
@thesamesam 

Hi Sam,

It might be worth commenting this here.

Here's an overview of the branch:
```
$ git log --oneline master..fgets 
cd64d530 (HEAD -> fgets, gh/fgets) lib/, po/: Remove fgetsx() and fputsx()
941c951c lib/, src/: Remove useless casts in fgets(3)
4d66df60 lib/, src/: Consistently use NULL with fgets(3)
d0cdc9c8 lib/gshadow.c: endsgent(): Invert logic to reduce indentation
a719bc39 (gh/string, string) lib/fields.c: Remove dead code
b71d2236 contrib/, lib/, src/: Use consistent style using strchr(3) in conditionals
44df1083 lib/, src/: Use str[n]cmp(3) instead of explicit byte comparisons
d3350b67 contrib/, lib/, src/: Use strchr(3) to compact comparisons
9a6e3284 lib/loginprompt.c: login_prompt(): Use strtcpy() instead of its pattern
db93a222 lib/, src/: Use strsep(3) instead of strtok(3)
376e4166 src/suauth.c: check_su_auth(): Use pointers to simplify
ae1c284e src/suauth.c: check_su_auth(): Use strspn(3) instead of its pattern
93362cc3 lib/gshadow.c: endsgent(): Remove dead assignment
5a46723a lib/port.c: portcmp(): Use strcmp(3) instead of its pattern
7442a127 lib/, src/: Use stpspn() instead of its pattern
```

This `fgets` branch is really only
```
$ git log --oneline string..fgets 
cd64d530 (HEAD -> fgets, gh/fgets) lib/, po/: Remove fgetsx() and fputsx()
941c951c lib/, src/: Remove useless casts in fgets(3)
4d66df60 lib/, src/: Consistently use NULL with fgets(3)
d0cdc9c8 lib/gshadow.c: endsgent(): Invert logic to reduce indentation
```
but I based it on `string`, which corresponds to PR <https://github.com/shadow-maint/shadow/pull/1048>.

Here's why.  Basically, according to the bug report that it closes, the PR is about replacing fgetsx => fgets.  Here are two of the calls that we replace:

```c
$ grep -rnC4 fgetsx lib/gshadow.c
154-	if (NULL == fp) {
155-		return NULL;
156-	}
157-
158:	if (fgetsx(buf, buflen, fp) == buf) {
159-		while (   ((cp = strrchr (buf, '\n')) == NULL)
160-		       && (feof (fp) == 0)) {
161-			size_t len;
162-
--
167-			buf = cp;
168-			buflen *= 2;
169-
170-			len = strlen (buf);
171:			if (fgetsx (&buf[len],
172-			            (int) (buflen - len),
173-			            fp) != &buf[len]) {
174-				return NULL;
175-			}
```

I want to fully understand what the code is doing before doing the change, which makes me more confident that the change is correct.  To that end, I need to reduce superfluous complexity.

In the code above, you may notice a dead assignment:
```c
159-		while (   ((cp = strrchr (buf, '\n')) == NULL)
```
I already had removed that dead assignment in <https://github.com/shadow-maint/shadow/pull/1048> in commit <https://github.com/shadow-maint/shadow/pull/1048/commits/93362cc39a78a12aa5732a83e28fddd5891bc63d>. That's the only reason why I decided to base it on that PR.

Now let's see why this PR has 4 new commits, instead of 1.

Line
```c
158:	if (fgetsx(buf, buflen, fp) == buf) {
```
Has a conditional that spans the entire rest of the function.  Commit 1 turns this into the usual thing, which is an early return on error.

Then we get to
```c
171:			if (fgetsx (&buf[len],
172-			            (int) (buflen - len),
173-			            fp) != &buf[len]) {
```

Why do we have that cast?  It's nonsense.  I'll remove it.
And also `&buf[len]` is very brittle, and a small typo could add a bug here.  Let's use NULL there.

While doing those two changes, I'll check any other places that do similarly bad stuff, and replace it in the entire code base.  Then commit 4 does what we originally wanted.

I don't reason too much the first 3 commits, which you might find frustrating while reviewing the changes a posteriori.  I didn't because they're not the main purpose of the PR; they're just cosmetic changes in preparation for the PR.

Do you have any suggestions for making it easier for you to review the changes later?

Maybe I could add an explicit `[cosmetic]` in the second line of the commit message, to mark those as unimportant commits.

Still, I need to apply those cosmetic changes, or it would be much more difficult to me to reason about the important changes of the PR.

Reply:
And these cosmetic patches combined with the change fgetsx=>fgets have allowed me to see a transformation fgets=>getline too.  I probably wouldn't have been able to see it without the cosmetic changes.

Reply:
Did you consider my coccinelle suggestion?

Reply:
I don't know how to use it.  If you point me to some docs and/or show me a small shell example session showing how it's useful, I'll be happy to try it.

Reply:
Hmmm.  <https://coccinelle.gitlabpages.inria.fr/website/sp.html>.  Sounds like it could be interesting.

Reply:
That's why I mentioned it before ;)

I think it's both a good fit for the work you're interested in _and_ eases review of it.

Reply:
I'm trying to develop a coccinelle script that would do something similar to <https://github.com/shadow-maint/shadow/pull/1056/commits/941c951cba51d8b050ed82854276054f45485ddc>

So far, I've written 
```diff
$ cat fgets_cast.sp 
@@
expression a, b, c;
type t;

@@

- fgets(a, (t) (b), c)
+ fgets(a, b, c)

@@
expression a, b, c;
type t;

@@

- fgets(a, (t) b, c)
+ fgets(a, b, c)

@@
expression a, b, c;
type t;
@@

- fgetsx(a, (t) (b), c)
+ fgetsx(a, b, c)

@@
expression a, b, c;
type t;
@@

- fgetsx(a, (t) b, c)
+ fgetsx(a, b, c)

@@
expression a, b, c, p;
type t;
@@

- p->fgets(a, (t) (b), c)
+ p->fgets(a, b, c)

@@
expression a, b, c, p;
type t;
@@

- p->fgets(a, (t) b, c)
+ p->fgets(a, b, c)
```

but it results in false positives, such as 
```diff
diff --git i/lib/shadow.c w/lib/shadow.c
index 44436836..5759f259 100644
--- i/lib/shadow.c
+++ w/lib/shadow.c
@@ -202,7 +202,7 @@ struct spwd *fgetspent (FILE * fp)
                return (0);
        }
 
-       if (fgets (buf, sizeof buf, fp) != NULL)
+       if (fgets(buf, sizeof buf, fp) != NULL)
        {
                stpsep(buf, "\n");
                return my_sgetspent (buf);
```

How is that matching `(t)`, if there are no casts?